software/cm/src/main/java/org/apache/brooklyn/entity/cm/ansible/AnsibleLifecycleEffectorTasks.java

```diff
@@ -0,0 +1,203 @@
+import org.apache.brooklyn.feed.ssh.SshPollConfig;
+import org.apache.brooklyn.location.ssh.SshMachineLocation;
+import org.apache.brooklyn.util.core.task.DynamicTasks;
+import org.apache.brooklyn.util.guava.Maybe;
+import org.apache.brooklyn.util.net.Urls;
+import org.apache.brooklyn.util.text.Strings;
+import org.apache.brooklyn.util.time.Duration;
+import org.apache.brooklyn.util.time.Time;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class AnsibleLifecycleEffectorTasks extends MachineLifecycleEffectorTasks implements AnsibleConfig {
+
+    private static final Logger LOG = LoggerFactory.getLogger(AnsibleLifecycleEffectorTasks.class);
+
+    protected String serviceName;
+    protected SshFeed serviceSshFeed;
+
+    protected Object extraVars;
+    protected String baseDir;
+    protected String runDir;
+
+    public AnsibleLifecycleEffectorTasks() {
+    }
+
+    public String getServiceName() {
+        if (serviceName!=null) return serviceName;
+        return serviceName = entity().config().get(AnsibleConfig.SERVICE_NAME);
+    }
+
+    public Object getExtraVars() {
+        if (extraVars != null) return extraVars;
+        return extraVars = entity().config().get(ANSIBLE_VARS);
+    }
+
+    public String getBaseDir() {
+        if (null != baseDir) return baseDir;
+        return baseDir = MachineLifecycleEffectorTasks.resolveOnBoxDir(entity(),
+                Machines.findUniqueMachineLocation(entity().getLocations(), SshMachineLocation.class).get());
+    }
+
+    public String getRunDir() {
+        if (null != runDir) return runDir;
+        return runDir = Urls.mergePaths(getBaseDir(), "apps/"+entity().getApplicationId()+"/ansible/playbooks/"
+            +entity().getEntityType().getSimpleName()+"_"+entity().getId());
+    }
+
+    @Override
+    public void attachLifecycleEffectors(Entity entity) {
+        if (getServiceName()==null && getClass().equals(AnsibleLifecycleEffectorTasks.class)) {
+            // warn on incorrect usage
+            LOG.warn("Uses of "+getClass()+" must define a PID file or a service name (or subclass and override {start,stop} methods as per javadoc) " +
+                    "in order for check-running and stop to work");
+        }
+        super.attachLifecycleEffectors(entity);
+    }
+
+    @Override
+    protected String startProcessesAtMachine(Supplier<MachineLocation> machineS) {
+        startWithAnsibleAsync();
+
+        return "ansible start tasks submitted";
+    }
+
+    protected String getPlaybookName() {
+        return entity().config().get(ANSIBLE_PLAYBOOK);
+    }
+
+    protected void startWithAnsibleAsync() {
+        String installDir = Urls.mergePaths(getBaseDir(), "installs/ansible");
+
+        String playbookUrl = entity().config().get(ANSIBLE_PLAYBOOK_URL);
+        String playbookYaml = entity().config().get(ANSIBLE_PLAYBOOK_YAML);
+
+        if (Strings.isNonBlank(playbookUrl) && Strings.isNonBlank(playbookYaml)) {
```

Comment:
Worth also adding a check for the case where both of them are blank, as otherwise the statements just below will run ansible without having established a playbook by either means.